specs/core/0_beacon-chain.md

```diff
@@ -1592,41 +1614,38 @@ _Note_: If there are skipped slots between a block and its parent block, run the
 
 ### Per-slot processing
 
-Below are the processing steps that happen at every slot.
-
-#### Slot
+Below are the processing steps that happen at every `slot >= GENESIS_SLOT`.
 
+* Set `state.latest_state_roots[state.slot % SLOTS_PER_BATCHING] = hash_tree_root(state)`.
+* Set `state.latest_block_roots[state.slot % SLOTS_PER_BATCHING] = get_block_root(state, state.slot - 1)`.
 * Set `state.slot += 1`.
-
-#### Block roots
-
-* Let `previous_block_root` be the `hash_tree_root` of the previous beacon block processed in the chain.
-* Set `state.latest_block_roots[(state.slot - 1) % LATEST_BLOCK_ROOTS_LENGTH] = previous_block_root`.
-* If `state.slot % LATEST_BLOCK_ROOTS_LENGTH == 0` append `merkle_root(state.latest_block_roots)` to `state.batched_block_roots`.
+* If `state.slot % SLOTS_PER_BATCHING == 0` append `merkle_root(state.latest_block_roots + state.latest_state_roots)` to `state.historical_batchings`.
 
 ### Per-block processing
 
-Below are the processing steps that happen at every `block`.
+Below are the processing steps that happen at every `block` except the genesis block.
 
-#### Slot
+#### Block header
 
 * Verify that `block.slot == state.slot`.
-
-#### Block signature
-
+* Let `previous_block_header = state.latest_block_header`.
+* Set `previous_block_header.state_root = get_state_root(state, state.slot - 1)`.
```

Comment:
This will not rebuild the "parent block header" in all cases (skipped slots). It will rebuild some weird intermediate made up block header because the way state is being pulled here. My first guess on fixing this would be to use the slot from the header rather than just the previous slot. Dumping this thought here and thinking about it tomorrow.

```suggestion
* Set `previous_block_header.state_root = get_state_root(state, previous_block_header.slot)`.
```

Reply:
Well spotted! Unfortunately `get_state_root` won't work when we have too many empty slots in a row. My suggestion would be to build the full `state.latest_block_header` (with `state_root`) on the next slot. Will fix soon :)